spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java

```diff
@@ -529,47 +527,63 @@ public final Connection createConnection() throws AmqpException {
 		Assert.state(!this.stopped, "The ApplicationContext is closed and the ConnectionFactory can no longer create connections.");
 		synchronized (this.connectionMonitor) {
 			if (this.cacheMode == CacheMode.CHANNEL) {
-				if (this.connection == null) {
-					this.connection = new ChannelCachingConnectionProxy(super.createBareConnection());
+				if (this.connection.target == null) {
+					this.connection.target = super.createBareConnection();
 					// invoke the listener *after* this.connection is assigned
+					if (!this.checkoutPermits.containsKey(this.connection)) {
+						this.checkoutPermits.put(this.connection, new Semaphore(this.channelCacheSize));
+					}
 					getConnectionListener().onCreate(this.connection);
-					this.checkoutPermits.put(this.connection, new Semaphore(this.channelCacheSize));
 				}
 				return this.connection;
 			}
 			else if (this.cacheMode == CacheMode.CONNECTION) {
 				ChannelCachingConnectionProxy connection = null;
-				while (connection == null && !this.idleConnections.isEmpty()) {
+				ChannelCachingConnectionProxy lastIdle = this.idleConnections.peekLast();
+				while (connection == null) {
 					connection = this.idleConnections.poll();
 					if (connection != null) {
 						if (!connection.isOpen()) {
 							if (logger.isDebugEnabled()) {
-								logger.debug("Removing closed connection '" + connection + "'");
+								logger.debug("Skipping closed connection '" + connection + "'");
 							}
 							connection.notifyCloseIfNecessary();
-							this.openConnections.remove(connection);
-							this.openConnectionNonTransactionalChannels.remove(connection);
-							this.openConnectionTransactionalChannels.remove(connection);
-							this.checkoutPermits.remove(connection);
+							this.idleConnections.addLast(connection);
+							if (connection.equals(lastIdle)) {
+								// all of the idle connections are closed.
+								connection = this.idleConnections.poll();
+								break;
+							}
 							connection = null;
```

Comment:
Looks like we have to add
```
this.checkoutPermits.remove(connection);
```
before this nulling.
Otherwise I see you re-new the `connection` in the next `if` and put it into the the `this.checkoutPermits`.

I can't catch that case with tests, so looks like I'm missing some very smart logic here around.

Would you mind elaborating more?

Thanks